zip-0222.rst

```diff
@@ -0,0 +1,314 @@
+
+TZE Input Encoding:
+
+====================== ==================================== ===============
+Field                  Description                          Type           
+====================== ==================================== ===============
+``prevout_hash``       hash of previous txid                ``uint256``    
+``prevout_in``         index into previous txn's outputs    ``uint32``
+``witness``            witness for prevout's precondition   ``tze_data``
+====================== ==================================== ===============
+
+
+TZE Output Encoding:
+
+====================== ==================================== ===============
+Field                  Description                          Type           
+====================== ==================================== ===============
+``amount``             spendable amount, in zatoshi         ``int64``    
+``precondition``       precondition encumbering ``amount``  ``tze_data``
+====================== ==================================== ===============
+
+
+Consensus rules
+---------------
+
+Once the TODO network upgrade activates, the following new consensus rules are enforced:
+
+- For each ``(outpoint, witness)`` pair in ``tze_inputs``:
+
+  - ``outpoint`` MUST reference a precondition of the same type and mode in an already-mined
+    transaction.
+  - ``tze_verify(mode, precondition, witness, context)`` MUST return ``true``.
+
+- If a transaction has non-empty ``tze_inputs`` and non-empty ``tze_outputs``, then every
+  element in both fields MUST have the same ``type``.
+
+- Non-coinbase transactions MUST have at least one of the following:
+  - nonempty transparent inputs
+  - nonempty shielded inputs
+  - nonempty ``tze_inputs``
+
+- Transactions MUST have at least one of the following:
+  - nonempty transparent outputs
+  - nonempty shielded outputs
+  - nonempty ``tze_outputs``
+
+- All ``amount`` field values of ``tze_output`` records MUST be nonnegative and less than
+  MAX_MONEY.
+
+- The sum of amounts associated with all transparent, shielded, and TZE outputs within 
+  a transaction MUST NOT exceed the sum of the amounts associated with transaction inputs.
+
+Changes to signatures over transaction digests
+----------------------------------------------
+
+The newly added parts of the transaction will be included in transaction 
+digests for signatures. See [#zip-0244]_ for the specification of the new
+digests.
+
+Rationale
+=========
+
+Transactions that have both TZE inputs and outputs are required to use a single extension
+type, in order to prevent cross-protocol attacks. The downside is that this prevents all
+TZE-encumbered value from being spent directly into a different TZE type; the value needs
+to go through a regular address in between. This restriction might be relaxed in future
+ZIPs for specific combinations of ``(type, mode)`` pairs that have been analyzed for
+cross-protocol attacks, but we opt here for a fail-safe default behaviour.
```

Comment:
In a second paragraph, note that we don't prevent transactions from having TZE inputs with different modes as long as there are no TZE outputs, and vice versa; this should be low-risk because any horizontal interaction would require a TZE input or output to access an adjacent TZE input or output via its context, which would be explicit in the extension's ZIP and therefore already analysed.

Reply:
Just modes, or both modes and extensions? The latter should be fine, right?

Reply:
@str4d I've updated this, please check that I've got it right.